1/main.rs

```diff
@@ -0,0 +1,34 @@
+// Parse standard in as a sequence of integers.
+fn input_ints() -> Vec<i32> {
+	use std::io::prelude::*;
+	std::io::stdin().lock().lines().map(
+		|line| line.unwrap().trim().parse::<i32>().unwrap()
+	).collect()
+}
+
+// Sum inputs in windows of 3.
+fn boxcar_sum(signal : Vec<i32>) -> Vec<i32> {
+    let mut result = Vec::new();
+    for n in 0..(signal.len() - 2) {
```

Comment:
Can I do this as a single expression? In python I could do something like this

range( 0, signal.len() - 2 ).filter( lambda x: x >= 2 ).map( lambda idx: signal[idx-2] + signal[idx-1] + signal[idx] )

but I can't find an equivalent expression in rust

Reply:
yep! it's actually also called `.map` in rust. here's an [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d65c6af83877cc45b9bb265f07b67998)

`x..y` is desugaring to a [Range](https://doc.rust-lang.org/std/ops/struct.Range.html) which implements [Iterator](https://doc.rust-lang.org/std/ops/struct.Range.html#impl-Iterator) which has [`.map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map) (and many other useful) methods.

Iterators in rust are sort of like the c++ ones (but with more borrow checking). 

Generally when doing data transforms you will have 3 rough steps. 
1. convert your vec or slice into an iterator with `.iter()` or `.into_iter` (depending on whether you want to borrow or own the inner elements).
2. call a bunch of iterator methods chained together
3. collect the data back into a vec with `.collect::<Vec<_>>()` (you have to tell the compiler what container you want to collect back into). alternatively you can collapse all of the items back down to one with things like `.sum()` and `.fold()`.